### ISSUE TITLE
Regression in `arainko/ducktape` under `-Xcheck-macros` when comparing field types

### ISSUE THREAD

**Issue by WojciechMazur** [lampepfl/dotty#18754]
Open CB failure found in @arainko /ducktape under `-Xcheck-macros`. [Builds logs](https://github.com/VirtusLab/community-build3/actions/runs/6593524845/job/17917573705)

The issue is only present when checking types of nested members (`contact.phoneNo`)

## Compiler version
3.4.0-RC1-bin-20231023-44a537b-NIGHTLY
Works in 3.3.1

## Minimized code
No minimization yet, reproducer: 

```Scala
//> using option -Xcheck-macros
//> using dep "io.github.arainko::ducktape:0.1.11"
import io.github.arainko.ducktape.*

final case class Primitive(contact: PrimitiveContactInfo)
final case class PrimitiveContactInfo(phoneNo: String)

final case class Complex(contact: ComplexContactInfo)
final case class ComplexContactInfo(phoneNo: PhoneNumber)
final case class PhoneNumber(value: String) extends AnyVal

def Test() = {
  given Transformer[String, PhoneNumber] = str => PhoneNumber(str + "-LOCAL")
  
  val primitive: Primitive = ???
  val _ = primitive.into[Complex].transform()
}

```

## Output

```scala
-- Error: /Users/wmazur/projects/sandbox/main.scala:16:43 ----------------------
16 |  val _ = primitive.into[Complex].transform()
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |Malformed tree was found while expanding macro with -Xcheck-macros.
   |               |The tree does not conform to the compiler's tree invariants.
   |               |
   |               |Macro was:
   |               |scala.quoted.runtime.Expr.splice[Complex](((evidence$7: scala.quoted.Quotes) ?=> io.github.arainko.ducktape.internal.macros.Transformations.inline$transformConfiguredMacro[Primitive, Complex](scala.quoted.runtime.Expr.quote[Primitive](source$proxy1).apply(using evidence$7), scala.quoted.runtime.Expr.quote[scala.collection.immutable.Seq[io.github.arainko.ducktape.BuilderConfig$package.BuilderConfig[Primitive, Complex]]](: _*).apply(using evidence$7))(scala.quoted.Type.of[Primitive](evidence$7), scala.quoted.Type.of[Complex](evidence$7), evidence$7)))
   |               |
   |               |The macro returned:
   |               |new Complex(contact = {
   |  val LowPriorityTransformerInstances_this: io.github.arainko.ducktape.Transformer.type = io.github.arainko.ducktape.Transformer
   |  new ComplexContactInfo(phoneNo = given_Transformer_String_PhoneNumber.transform(source$proxy1.contact.phoneNo))
   |})
   |               |
   |               |Error:
   |               |assertion failed: Types differ
   |Original type : (source.phoneNo : String)
   |After checking: (source$proxy1.contact.phoneNo : String)
   |Original tree : source$proxy1.contact.phoneNo
   |After checking: source$proxy1.contact.phoneNo
   |Why different :
   |             Subtype trace:
   |  ==> (source$proxy1.contact.phoneNo : String)  <:  (source.phoneNo : String)
   |    ==> (source$proxy1.contact : PrimitiveContactInfo)  <:  (source : PrimitiveContactInfo)
   |      ==> PrimitiveContactInfo  <:  (source : PrimitiveContactInfo) (left is approximated)
   |      <== PrimitiveContactInfo  <:  (source : PrimitiveContactInfo) (left is approximated) = false
   |    <== (source$proxy1.contact : PrimitiveContactInfo)  <:  (source : PrimitiveContactInfo) = false
   |    ==> String  <:  (source.phoneNo : String) (left is approximated)
   |      ==> String  <:  (source.phoneNo : String) (left is approximated)
   |      <== String  <:  (source.phoneNo : String) (left is approximated) = false
   |    <== String  <:  (source.phoneNo : String) (left is approximated) = false
   |  <== (source$proxy1.contact.phoneNo : String)  <:  (source.phoneNo : String) = false
   |               |
   |stacktrace available when compiling with `-Ydebug`
   |               |
   |----------------------------------------------------------------------------
   |Inline stack trace
   |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   |This location contains code that was inlined from Transformations.scala:69
   |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
   |This location contains code that was inlined from Transformations.scala:69
    ----------------------------------------------------------------------------
1 error found
```

## Expectation
Should not raise an error

**Comment by WojciechMazur** [comment]
Project-wise bisect results: 
Last good release: 3.4.0-RC1-bin-20230726-97677cc-NIGHTLY
First bad release: 3.4.0-RC1-bin-20230727-ac69c66-NIGHTLY

The commit-based bisect failed due to missing sbt-bridge 
Candidate commits: 
ac69c66ec84171b13e86dd4b4197bd86fd930c62
d903d35eb0a697a8182e16e88cd28b840a70d8d9
cb73f0b35b239f626152cd260442cc6172cc122d
c04d2db43a23d18d895878eedb02784c2eabf830
989c55c83a282756595307515512b85ce73384a1  [bad commit]
4b67b34dcb05c5b5efd9548087607a5d73f772c4 [bad commit] possible regression introducer
182331b6dfec1fc987ef3c7a923d936a0d22936a [ good commit ]
56251079218efae3d8fc9dd618ad009e7a396a86
f22f9db9fc4c971ebc862521a22bcd4551f64d02 [problems with sbt-bridge]
101ce3afca2a0a6c41783fa7cbf1fdf1be2281c6
f5313ae761d8c086705e29d6d41296b6a1c58e4c
96ca0b6bc8504a26359e9fab897f2d79b861028f

**Comment by nicolasstucki** [comment]
We need a self-contained version of this code to make any progress on this issue.  @arainko could you help us with that?

**Comment by arainko** [comment]
Sure, I'll try to minimize this over the weekend

**Comment by arainko** [comment]
Ok, so my findings so far (I couldn't get a full minimization yet) - the error happens after a call to [LiftTransformation.liftTransformation](https://github.com/arainko/ducktape/blob/a1944cc0fbacea2af00deebecc07ca992c2934f9/ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/LiftTransformation.scala#L13) is made and then when the macro falls into the [liftDerivedTransformation(transformer, appliedTo)](https://github.com/arainko/ducktape/blob/a1944cc0fbacea2af00deebecc07ca992c2934f9/ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/LiftTransformation.scala#L113) case.

What this specific piece of code does:
1. It matches a specific call to a `.make` method on one of the specialized cases of a `Transformer` (`Transformer.ForProduct` in this case) and extracts:
  * The `ValDef` that  corresponds to the SAM lambda argument that creates that Transformer
  * All the definitions from `Inlined` nodes (the ones responsible for stuff like `source$proxy1` etc) that were encountered along the way
  * the method call to a constructor of the destination type
  * the terms that correspond to the arguments of that constructor
so for something like this
```scala
Transformer.ForProduct.make((p: Person) => new Person2(p.int, p.str))
```
We'll get `TransformerLambda.ForProduct( ValDef of 'p', <stuff from Inlined nodes>, 'new Person2', List(p.int, p.str))`

2. After that's done it calls [optimizeTransformerInvocation](https://github.com/arainko/ducktape/blob/a1944cc0fbacea2af00deebecc07ca992c2934f9/ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/LiftTransformation.scala#L122) and falls into the `TransformerLambda.ForProduct` case which will rewrite the tree of the original transformer to elide an instantiation of a new object by - so for example, this manually written out Transformer:
```scala
  final case class ToplevelPrimitive(value: ReproPrimitive)
  final case class ToplevelComplex(value: ReproComplex)

  final case class ReproPrimitive(name: String)
  final case class ReproComplex(name: Name)

  final case class Name(value: String)

  given Transformer[String, Name] = str => Name(str + "-LOCAL")

  val primitive: ToplevelPrimitive = ???

  inline def manualTransformer = Transformer.ForProduct.make[ToplevelPrimitive, ToplevelComplex](
    (source: ToplevelPrimitive) =>
      new ToplevelComplex(value = 
        new ReproComplex(
          name = given_Transformer_String_Name.transform(source.value.name)
        )  
      )
  ).transform(primitive)
```
will be rewritten to this:
```scala
(new ToplevelComplex(value = new ReproComplex(name = given_Transformer_String_Name.transform(primitive.value.name))): ToplevelComplex)
```
so it kind of 'lifts' the transformation from the Transformer's body by getting the RHS of the SAM lambda that creates it and replaces all of the references to the lambda parameter with the expression passed into `Transformer#transform` (so `primitive` in this case) and it also does this recursively for any invocations of `Transformer.ForProduct.make` that it encounters in the body of the toplevel transformer.

The error seems to occur when the body of `given_Transformer_String_Name.transform(source.value.name)` gets replaced with `given_Transformer_String_Name.transform(primitive.value.name)` aka when replacing arguments passed in to `Transformers` from implicit scope.

(I'm writing this in hopes that it gives an AHA moment to somebody :smile: )

I'll try to tackle this once again when I've got some free time.

**Comment by WojciechMazur** [comment]
Fixed in both LTS (3.3.3) and 3.nightly. Failing in 3.4.0, compiles in 3.4.1-RC1
